### PR TITLE
State comparison json

### PR DESCRIPTION
**Pull request by alonmuroch**
Introducing json state comparison per test.

**Goal**

- Readable post state 
- Comparable (via json diff) post state
- Easy to spot changes (json will change) 

**Overview**
- Post test the state is written into a per test file
- When test loads, it will override (for supported test types) the expected post root and state
- The above will be used to verify the test 

(Supports only msg processing tests for now, can be expanded later)

### CODE REVIEW

File: qbft/spectest/generate/main.go [hunk 1]
```diff
@@ -37,6 +41,59 @@ func main() {
...
+	}
+
+	if err := os.Mkdir(dir, 0700); err != nil {
+		panic(err.Error())
+	}
+}
+
+func writeJsonStateComparison(name, testType string, post interface{}) {
+	if post == nil { // If nil, test not supporting post state comparison yet
+		fmt.Printf("skipping state comparison json, not supported: %s\n", name)
```
Comment (GalRogozinski) [inline qbft/spectest/generate/main.go:74]: probably should use some logger lib

File: ssv/spectest/generate/main.go [hunk 2]
```diff
@@ -37,6 +41,60 @@ func main() {
...
+	}
+
+	if err := os.Mkdir(dir, 0700); err != nil {
+		panic(err.Error())
+	}
+}
+
+func writeJsonStateComparison(name, testType string, post interface{}) {
+	if post == nil { // If nil, test not supporting post state comparison yet
+		fmt.Printf("skipping state comparison json, not supported: %s\n", name)
```
Comment (GalRogozinski) [inline ssv/spectest/generate/main.go:75]: ditto

File: qbft/spectest/tests/controller_spectest.go [hunk 3]
```diff
@@ -180,3 +189,52 @@ func (test *ControllerSpecTest) runInstanceWithData(
...
+	for i, runData := range test.RunInstanceData {
+		runData.ControllerPostState = sc[i]
+
+		r, err := sc[i].GetRoot()
+		require.NoError(t, err)
+		runData.ControllerPostRoot = hex.EncodeToString(r[:])
+	}
+}
+
+func (test *ControllerSpecTest) GetPostState() (interface{}, error) {
```
Comment (GalRogozinski) [inline qbft/spectest/tests/controller_spectest.go:211]: if we change something in the run function we need to change it here as well..

File: qbft/spectest/tests/controller_spectest.go [hunk 4]
```diff
@@ -43,17 +47,14 @@ func (test *ControllerSpecTest) TestName() string {
...
 func (test *ControllerSpecTest) Run(t *testing.T) {
-	identifier := []byte{1, 2, 3, 4}
-	config := testingutils.TestingConfig(testingutils.Testing4SharesSet())
-	contr := testingutils.NewTestingQBFTController(
-		identifier[:],
-		testingutils.TestingShare(testingutils.Testing4SharesSet()),
-		config,
-	)
+	// temporary to override state comparisons from file not inputted one
+	test.overrideStateComparison(t)
```
Comment (GalRogozinski) [inline qbft/spectest/tests/controller_spectest.go:51]: why is this needed?
why did the roots change

File: qbft/spectest/generate/main.go (hunk 1 above)
Comment (alonmuroch) [reply qbft/spectest/generate/main.go:74]: change as you see fit (not sure we use any lib in tests)

File: qbft/spectest/tests/controller_spectest.go (hunk 3 above)
Comment (alonmuroch) [reply qbft/spectest/tests/controller_spectest.go:211]: yes

**Review by GalRogozinski — CHANGES_REQUESTED** [review]

File: qbft/spectest/tests/controller_spectest.go [hunk 5]
```diff
@@ -65,9 +66,19 @@ func (test *ControllerSpecTest) Run(t *testing.T) {
...
+		identifier[:],
+		testingutils.TestingShare(testingutils.Testing4SharesSet()),
+		config,
+	)
+}
+
 func (test *ControllerSpecTest) testTimer(
 	t *testing.T,
-	config *qbft.Config,
+	config qbft.IConfig,
```
[the remark this answers is not in the document]
Comment (alonmuroch) [reply qbft/spectest/tests/controller_spectest.go:81]: IConfig is a interface not a struct

File: qbft/spectest/tests/controller_spectest.go [hunk 6]
```diff
@@ -81,7 +92,7 @@ func (test *ControllerSpecTest) testTimer(
 func (test *ControllerSpecTest) testProcessMsg(
 	t *testing.T,
 	contr *qbft.Controller,
-	config *qbft.Config,
+	config qbft.IConfig,
```
[the remark this answers is not in the document]
Comment (alonmuroch) [reply qbft/spectest/tests/controller_spectest.go:95]: same as above

File: qbft/spectest/tests/controller_spectest.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply qbft/spectest/tests/controller_spectest.go:81]: oops didn't see the added `I` 😅

File: qbft/spectest/tests/controller_spectest.go [hunk 7]
```diff
@@ -180,3 +189,52 @@ func (test *ControllerSpecTest) runInstanceWithData(
 
 	return lastErr
 }
+
+func (test *ControllerSpecTest) overrideStateComparison(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alonmuroch) [reply qbft/spectest/tests/controller_spectest.go:193]: require hard coded root to be equal to loaded root

File: ssv/spectest/tests/msg_processing_spectest.go [hunk 8]
```diff
@@ -31,21 +35,9 @@ func (test *MsgProcessingSpecTest) TestName() string {
...
-		lastErr = v.StartDuty(test.Duty)
-	}
-	for _, msg := range test.Messages {
-		err := v.ProcessMessage(msg)
-		if err != nil {
-			lastErr = err
-		}
-	}
+// RunAsMultiTest runs the test as part of a MultiMsgProcessingSpecTest
+func (test *MsgProcessingSpecTest) RunAsMultiTest(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alonmuroch) [reply ssv/spectest/tests/msg_processing_spectest.go]: added comment

File: qbft/spectest/tests/controller/futuremsg/test.go [hunk 9]
```diff
@@ -66,3 +66,7 @@ func (test *ControllerSyncSpecTest) Run(t *testing.T) {
 		require.NoError(t, lastErr)
 	}
 }
+
+func (test *ControllerSyncSpecTest) GetPostState() (interface{}, error) {
+	return nil, nil
```
Comment (GalRogozinski) [inline qbft/spectest/tests/controller/futuremsg/test.go:71]: even though there is a post state here, we can generate it once we start working on the respective tests

File: qbft/spectest/generate/main.go (hunk 1 above)
Comment (y0sher) [reply qbft/spectest/generate/main.go:74]: go's built in logger `import "log"` should be enough for this.
If we want to think about a library-wide logging strategy and library to use, it's for sure out of this PR scope.

File: qbft/spectest/generate/main.go [hunk 10]
```diff
@@ -37,6 +41,59 @@ func main() {
...
+		writeJsonStateComparison(test.TestName(), reflect.TypeOf(test).String(), post)
+	}
+}
+
+func clearStateComparisonFolder() {
+	_, basedir, _, ok := runtime.Caller(0)
+	if !ok {
+		panic("no caller info")
+	}
+	dir := filepath.Join(strings.TrimSuffix(basedir, "main.go"), "state_comparison")
```
Comment (y0sher) [inline qbft/spectest/generate/main.go:61]: `main.go` seems redundant here, `filepath.Join(basedir, "state_comparison")` would give the same result

File: qbft/spectest/generate/main.go [hunk 11]
```diff
@@ -16,9 +16,13 @@ import (
 //go:generate go run main.go
 
 func main() {
+	clearStateComparisonFolder()
```
Comment (y0sher) [inline qbft/spectest/generate/main.go:19]: do we want to make clear the folder in case of panic in the middle? for example if we successfully wrote tests, but could not write state comparison, should we erase all in this case?
this will require `defer recover` or replacing the panics and tracing the error.

File: ssv/spectest/tests/valcheck/test.go [hunk 12]
```diff
@@ -62,6 +62,10 @@ func (test *SpecTest) valCheckF(signer types.BeaconSigner) qbft.ProposedValueChe
 	}
 }
 
+func (tests *SpecTest) GetPostState() (interface{}, error) {
+	return nil, nil
```
Comment (y0sher) [inline ssv/spectest/tests/valcheck/test.go:66]: these tests don't have any post state?

File: qbft/spectest/generate/state_comparison/*tests.ControllerSpecTest/qbft controller decide duplicate msg.json [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+												],
+												"Quorum": 3,
+												"PartialQuorum": 2,
+												"DomainType": [
+														0,
+														0,
+														3,
+														1
+												],
+												"FeeRecipientAddress": [
```
Comment (y0sher) [inline qbft/spectest/generate/state_comparison/*tests.ControllerSpecTest/qbft controller decide duplicate msg.json:38]: don't we want to encode this into hex string as well? just like other addresses we have or public keys. it will probably be more readable.

**Review by y0sher — APPROVED** [review]

File: qbft/spectest/generate/main.go (hunk 11 above)
Comment (GalRogozinski) [reply qbft/spectest/generate/main.go:19]: I personally don't see it as something serious.
The jsons will be in git and we will never merge in something that panics

File: ssv/spectest/tests/valcheck/test.go (hunk 12 above)
Comment (GalRogozinski) [reply ssv/spectest/tests/valcheck/test.go:66]: this specifically no

There are some places though that we return nil even though there is a post state.
But in those places we can implement this function as we touch the tests later.

Personally for me, this PR is more about getting the infra right

Comment (GalRogozinski) [reply ssv/spectest/tests/valcheck/test.go:66]: This specifically no.
For others we will add later as needed.

**Review by GalRogozinski — APPROVED** [review]
